DataServer/server.py

```diff
@@ -286,29 +297,74 @@ def __getRevision(self) -> str:
             git_rev_parse += ' dirty'
         return git_rev_parse
 
-    def __init__(self, config_file: str) -> None:
+    def __init__(self, config_file: str, report_outage: Optional[Callable[[devices.Device], None]] = None) -> None:
         self._log = logging.getLogger(self.__class__.__name__)
         self._log.info(f"Starting ASM Data Server v{DataServer.__version__}, {self.__getRevision()}")
         self.config = ServerConfig(config_file)
         devices_file = os.path.join(self.config.data_dir, 'devices.yaml')
         self.device_tree = devices.DeviceTree(devices_file)
         self.hostname = ''
         self.__client_queues: List[ClientHandler] = []
+        self._report_outage = report_outage
+
+        for device in self.device_tree.getDevices():
+            self.setup_device(device)
 
     async def run(self):
         self._log.info(f'Connecting to {self.hostname}:{self.config.port}')
+
         server = await asyncio.start_server(self.client_thread, self.hostname,
                                             self.config.port)
+        outage_detector = asyncio.create_task(self.start_outage_detector())
         async with server:
             await server.serve_forever()
 
+        outage_detector.cancel()
+
     async def client_thread(self, reader: StreamReader, writer: StreamWriter):
         client = ClientHandler(device_tree=self.device_tree, reader=reader,
-                               writer=writer, config=self.config)
+                               writer=writer, config=self.config,
+                               setup_device=self.setup_device)
         self.__client_queues.append(client)
         await client.run()
         self.__client_queues.remove(client)
 
+    def on_device_timeout(self, device: devices.Device):
+        if self._report_outage:
+            self._report_outage(device)
+        else:
+            self._log.warn(f'Device {device.deviceID} went offline but had no report outage hook')
```

Comment:
It is probably okay to be verbose about the warning and log it even if there is outage callback.